### PR TITLE
[WIP] Improvement of testing routines

### LINKED ISSUE #64: Comprehensive testing of conservative regridding scheme (and other interface functions)
As part of the work in [CAM-ML](https://github.com/m2lines/CAM-ML) (specifically [#14](https://github.com/m2lines/CAM-ML/pull/14/commits)) there were a couple of bugfixes to the regridding scheme that have been propagated to this repository in #44 

As part of #44 @jatkinson1000 added tests for regridding and variable conversion.
However, these should be revisited and made more robust once #44 is merged to main.

Some ideas

- [ ] Revisit conservative regridding tests and 
  - [ ] Test conservation of mass in a column
    - [ ] For density 1.0
    - [ ] For more arbitrary density profile
  - [ ] Test for conversion between actual CAM and SAM grids
- [ ] Revise tests linear interpolation scheme as appropriate
- [ ] Refine the tests for variable conversion
- [ ] Write tests for any other routines as would be useful
- [ ] Run tests as part of the CI process?

### PR DESCRIPTION
**Pull request by AdelekeBankole**
This PR will close #64

### CODE REVIEW

**Review by jatkinson1000 — COMMENTED** [review]
Thanks @AdelekeBankole This looks great, thank you!

We have just run these locally and everything is passing.
Just left a couple of comments for things to check or suggestions to make things a little clearer.

File: YOG_convection/SAM_consts.F90 [hunk 1]
```diff
@@ -85,7 +85,10 @@ module SAM_consts_mod
   ! ---------------------------
   integer, parameter :: input_ver_dim = 30
       !! The number of cells in a SAM atmospheric column on which the neural net was trained
-  
+  integer, parameter :: num_sam_cells = 30
+      !! number of SAM cells
```
[the remark this answers is not in the document]
Comment (AdelekeBankole) [reply YOG_convection/SAM_consts.F90:89]: Resolved in other file `nn_convection_flux.F90` as well

File: tests/test_CAM_interface/test_cam_interface.F90 [hunk 2]
```diff
@@ -14,68 +16,75 @@ module cam_tests
...
-      real(dp), dimension(4, 48) :: p_cam, p_int_cam
-      real(dp), dimension(4, 48) :: var_cam
-      real(dp), dimension(4) :: var_cam_surface
-      real(dp), dimension(4) :: ps_cam
-      real(dp), dimension(4, 48) :: var_sam, var_sam_exp
+      real(dp), dimension(num_cols, nrf) :: p_cam, p_int_cam
+      real(dp), dimension(num_cols, nrf) :: var_cam
+      real(dp), dimension(num_cols) :: var_cam_surface
+      real(dp), dimension(num_cols) :: ps_cam
+      real(dp), dimension(num_cols, nrf) :: var_sam, var_sam_exp
```
[the remark this answers is not in the document]
Comment (AdelekeBankole) [reply tests/test_CAM_interface/test_cam_interface.F90:38]: I think they should be size `nrf`

**Review by jatkinson1000 — APPROVED** [review]
Thanks @AdelekeBankole 
I still have a query around the change of array size, but do not feel it is blocking provided we log this in the issues.

This is great stuff and makes the code more robust, thank you.

File: tests/test_CAM_interface/test_cam_interface.F90 [hunk 3]
```diff
@@ -410,10 +606,10 @@ program run_cam_tests
 
   implicit none
 
-  real(dp), dimension(4) :: a = [1.0, 2.0, 3.0, 4.0]
+  real(dp), dimension(num_cols) :: a = [1.0, 2.0, 3.0, 4.0]
   real(dp), dimension(5) :: b = [1.0, 2.0, 3.0, 4.0, 6.0]
-  integer, dimension(4) :: aa = [1, 2, 3, 4]
-  integer, dimension(4) :: bb = [1, 2, 3, 4]
+  integer, dimension(num_cols) :: aa = [1, 2, 3, 4]
+  integer, dimension(num_cols) :: bb = [1, 2, 3, 4]
```
[the remark this answers is not in the document]
Comment (AdelekeBankole) [reply tests/test_CAM_interface/test_cam_interface.F90]: They are not used, they are dead lines of array, they were definitely carried over from older version. They have been cleaned out.

**Pull request merged**